ARROW-15892: Dataset APIs require s3:ListBucket Permissions

sanjibansg opened the pull request:
This PR adds a boolean flag which can be used to avoid creating directories with the Dataset API for filesystems like s3.

pitrou commented:
@sanjibansg I'm not sure it will be easy to create an automated S3 test with limited permissions, so we may simply punt on that. @westonpace What do you think?

westonpace commented:
@pitrou We have precedence for it here: https://github.com/apache/arrow/blob/64560af62391cb8823b3af3ac60991c0de59bac3/python/pyarrow/tests/test_fs.py#L242 so I don't think it would be too difficult to add.

westonpace commented on python/pyarrow/tests/conftest.py:
```diff
@@ -311,3 +313,23 @@ def s3_server(s3_connection):
+
+
+@pytest.fixture(scope='session')
+def limited_s3_user(request, s3_server):
```
Why is this method here instead of `util.py`?

pitrou replied:
Also, why is this a fixture if it's invoked as a regular function?

sanjibansg replied:
I was trying to pass arguments in the fixture. See this https://stackoverflow.com/a/44701916
The fixture `limited_s3_user` can then be used with both the limited policy and the put only policy.

pitrou replied:
This doesn't answer the question: why do you want it to be a fixture?

sanjibansg replied:
Sorry, this fixture thing was a little confusing at first. I thought by decorating it as a fixture it should set up the s3 properly with the required permissions. But, that might be the reason of the failing CI. Updated the code where it is no longer a fixture, and moved it to util.py as well. Thanks for the suggestion!

westonpace commented on python/pyarrow/includes/libarrow_dataset.pxd:
```diff
@@ -18,6 +18,7 @@
 # distutils: language = c++
 
 from libcpp.unordered_map cimport unordered_map
+from libcpp cimport bool
```
So this works ([it creates an alias for `bint` named `bool`](https://github.com/cython/cython/blob/d2d2ea33a21e1915a52790ef64fdd1f28867854c/Cython/Includes/libcpp/__init__.pxd#L2)) but elsewhere we either use `bint` directly or call it `c_bool` with: `from libcpp cimport bool as c_bool`.  Lets stick with one of the two existing approaches instead of adding a third.

sanjibansg replied:
Made the change, thanks!

westonpace commented on python/pyarrow/tests/test_dataset.py:
```diff
@@ -4286,6 +4287,55 @@ def test_write_dataset_s3(s3_example_simple):
+        pa.array(np.repeat(['a', 'b'], 10))],
+        names=["f1", "f2", "part"]
+    )
+    # writing with filesystem object with create_dir flag set to false
```
Can you add a check to make sure that writing the dataset with `create_dir=True` fails?  It doesn't really give us any coverage of our code but it allows us to make sure we've configured the limited server correctly so the test is more robust.

sanjibansg replied:
Made the change.

westonpace reviewed (changes requested):
There are some lint errors:

```
/arrow/python/pyarrow/tests/test_fs.py:23:1: F401 're' imported but unused
/arrow/python/pyarrow/tests/test_fs.py:24:1: F401 'subprocess' imported but unused
/arrow/python/pyarrow/tests/test_fs.py:25:1: F401 'sys' imported but unused
/arrow/python/pyarrow/tests/test_fs.py:26:1: F401 'time' imported but unused
/arrow/python/pyarrow/tests/test_dataset.py:28:1: F401 'venv.create' imported but unused
```

pitrou commented on cpp/src/arrow/dataset/file_base.h:
```diff
@@ -387,6 +387,8 @@ struct ARROW_DS_EXPORT FileSystemDatasetWriteOptions {
   /// Controls what happens if an output directory already exists.
   ExistingDataBehavior existing_data_behavior = ExistingDataBehavior::kError;
 
+  bool create_dir = true;
```
Can you add a docstring for this?

sanjibansg replied:
Added docstring.

pitrou commented on cpp/src/arrow/dataset/dataset_writer.cc:
```diff
@@ -476,7 +478,8 @@ class DatasetWriter::DatasetWriterImpl : public util::AsyncDestroyable {
   Future<> DoWriteRecordBatch(std::shared_ptr<RecordBatch> batch,
-                              const std::string& directory, const std::string& prefix) {
+                              const std::string& directory, const std::string& prefix,
+                              const bool& create_dir) {
```
No need to pass primitive types by reference.
```suggestion
                              bool create_dir) {
```

pitrou replied:
Also, I may be misunderstanding, but this parameter doesn't seem actually used?

sanjibansg replied:
yes, sorry we don't need to pass it by reference. Yes, the parameter is actually needed. Thanks for pointing that out, thanks!

pitrou commented on python/pyarrow/dataset.py:
```diff
@@ -753,7 +753,7 @@ def write_dataset(data, base_dir, basename_template=None, format=None,
                   max_rows_per_file=None, min_rows_per_group=None,
                   max_rows_per_group=None, file_visitor=None,
-                  existing_data_behavior='error'):
+                  existing_data_behavior='error', create_dir=True):
```
Can you add the corresponding parameter description in the docstring below?

sanjibansg replied:
Added the parameter desc, thanks!

pitrou commented on python/pyarrow/tests/test_dataset.py:
```diff
@@ -4286,6 +4287,55 @@ def test_write_dataset_s3(s3_example_simple):
+
+@pytest.mark.parquet
+@pytest.mark.s3
+def test_write_dataset_s3_put_only(s3_server, limited_s3_user):
```
Can you add a comment explaining what this test is for, and referencing the JIRA issue number?

sanjibansg replied:
Added the comment.

pitrou commented on cpp/src/arrow/dataset/file_base.h:
```diff
@@ -387,6 +387,9 @@ struct ARROW_DS_EXPORT FileSystemDatasetWriteOptions {
   /// Controls what happens if an output directory already exists.
   ExistingDataBehavior existing_data_behavior = ExistingDataBehavior::kError;
 
+  /// Flag to restrict or allow creating directory for writing the dataset.
```
```suggestion
  /// Whether to attempt creating the dataset directory.
  /// This can be set to false to work around limited permissions on some filesystems.
```

sanjibansg replied:
Made the change, thanks!

pitrou commented on python/pyarrow/tests/util.py:
```diff
@@ -349,3 +351,102 @@ def signal_wakeup_fd(*, warn_on_full_buffer=False):
+        return False
+
+
+def limited_s3_user(s3_server, policy):
```
Since this is not a fixture anymore, can you merge this with `_configure_limited_user` and name the resulting function e.g. `_configure_s3_limited_user`?

sanjibansg replied:
Made the change.

pitrou commented on python/pyarrow/tests/util.py:
```diff
@@ -30,6 +30,8 @@
 import string
 import subprocess
 import sys
+import re
```
Nit, but please let's keep stdlib imports in alphabetical order.

sanjibansg replied:
Made the change.

pitrou commented on python/pyarrow/tests/test_dataset.py:
```diff
@@ -4286,6 +4286,66 @@ def test_write_dataset_s3(s3_example_simple):
+    ds.write_dataset(
+        table, "existing-bucket", filesystem=fs,
+        format="feather", create_dir=False
+    )
```
This test doesn't do any partitioning?

sanjibansg replied:
I don't think we can use Hive or Directory partitioning here, as we are keeping the `create_dir` as `False`. However, we can use Filename Partitioning. But, while doing that, I noticed there is probably some issues with Filename Partitioning, where the values of the field on which the partitioning is done is not retrieved correctly. All those values were returned as `null`. I am investigating the cause, if it's a simple fix, I can then maybe push that in this PR.

pitrou replied:
Hmm, if we can't do anything else than filename partitioning then is it worth fixing this issue? @westonpace What do you think?

pitrou replied:
As for the filename partioning bug, better to file a separate JIRA IMHO.

westonpace replied:
The original ask was for a case that didn't have any partitioning.  If an S3 user has a partitioning then they shouldn't run into the original issue because all CreateDir calls will be for bucket + path.  So this flag is only to enable the very specific case where partitioning is not used.

That being said, I think hive & directory partitioning should still work with `create_dir=False` if you were using s3 or a similar filesystem that did not require directories to be created in advance.  So I think it still has some general value if users wanted to avoid creating marker directories in their S3 repository.

We could also solve this problem by modifying s3fs.cc so that it didn't try and create the bucket if it already existed.  This would add a `BucketExists` call to the path [here](https://github.com/apache/arrow/blob/c73870acdc775c43b00c1816f9c17e78036c8025/cpp/src/arrow/filesystem/s3fs.cc#L2331).

pitrou replied:
I see. So should we add a test to check that it works anyway?

sanjibansg replied:
I have modified the test to use Hive partitioning with the `existing_data_behavior` set to `'overwrite_or_ignore'`. But, not very sure, why the Python CI tests are failing with the ChildProcessError.

westonpace commented on cpp/src/arrow/dataset/file_base.h:
```diff
@@ -387,6 +387,10 @@ struct ARROW_DS_EXPORT FileSystemDatasetWriteOptions {
   /// Controls what happens if an output directory already exists.
   ExistingDataBehavior existing_data_behavior = ExistingDataBehavior::kError;
 
+  /// Whether to attempt creating the dataset directory.
+  /// This can be set to false to work around limited permissions on some filesystems.
+  bool create_dir = true;
```
```suggestion
  /// \brief If false the dataset writer will not create directories
  /// This is mainly intended for filesystems that do not require directories such as S3
  bool create_dir = true;
```

sanjibansg replied:
Made the change.

westonpace commented on python/pyarrow/dataset.py:
```diff
@@ -852,6 +852,8 @@ def file_visitor(written_file):
         dataset.  The first time each partition directory is encountered
         the entire directory will be deleted.  This allows you to overwrite
         old partitions completely.
+    create_dir : bool, default True
+        Flag to restrict or allow creating directory while writing a dataset.
```
```suggestion
    create_dir : bool, default True
        If False, directories will not be created.  This can be useful for filesystems
        that do not require directories.
```

sanjibansg replied:
Made the change.

kszucs commented:
@sanjibansg please rebase to make the builds pass

sanjibansg commented on python/pyarrow/tests/util.py:
```diff
@@ -350,3 +353,98 @@ def signal_wakeup_fd(*, warn_on_full_buffer=False):
+            return False
+        if not _ensure_minio_component_version('minio', 2021):
+            # minio version is too old for the capabilities we need
+            return False
```
Removed _ensure_minio_component_version.

sanjibansg commented on python/pyarrow/tests/util.py:
```diff
@@ -350,3 +353,98 @@ def signal_wakeup_fd(*, warn_on_full_buffer=False):
+        # The s3_server fixture starts the minio process but
+        # it takes a few moments for the process to become available
+        _wait_for_minio_startup(mcdir, address, access_key, secret_key)
+        # These commands create a limited user with a specific
```
Starting the try block from `_wait_for_minio_startup`. Made the change, thanks!

sanjibansg commented on python/pyarrow/tests/util.py:
```diff
@@ -350,3 +353,98 @@ def signal_wakeup_fd(*, warn_on_full_buffer=False):
+                        'myminio', 'no-create-buckets', 'user=limited')
+        _run_mc_command(mcdir, 'mb', 'myminio/existing-bucket')
+
+    except FileNotFoundError:
```
Sorry, I do not understand. Are you saying whether the `FileNotFoundError` will even be raised since we ran `_ensure_minio_component_version`?

pitrou replied:
Indeed, that's my question.

pitrou replied:
Also I now notice you've removed the calls to `_ensure_minio_component_version`, is that intended? @westonpace What is your take on this?

sanjibansg replied:
Oh, sorry, I misunderstood this comment
https://github.com/apache/arrow/pull/12701#discussion_r857420906

I am getting back the `_ensure_minio_component_version` function, and we can maybe remove the try-except block since the error should not be raised if `_ensure_minio_component_version` ran successfully.

pitrou replied:
Ok, can you explicitly skip the test when the Minio checks fail?

sanjibansg replied:
Skipping the test explicitly if the check fails.

pitrou commented:
Well, my bad: it seems the `FileNotFoundError` exception can still happen :-)
Also, the test needs skipping on Windows (see AppVeyor results: `mc` is something else on Windows)

sanjibansg commented:
> Well, my bad: it seems the `FileNotFoundError` exception can still happen :-) Also, the test needs skipping on Windows (see AppVeyor results: `mc` is something else on Windows)

Yes, the Windows issue should be fixed now, moved that line of code to the correct place. Modified the `_ensure_minio_component_version` to raise a `FileNotFoundError` if the correct version is not found, and then the test gets skipped.

pitrou commented:
@github-actions crossbow submit -g python
